File: samples/conference/registration-bc/src/main/java/com/microsoft/conference/registration/web/Controllers/RegistrationController.cs
using System;
using System.Linq;
using System.Threading.Tasks;
using System.Web.Mvc;
using Conference.Common;
using ECommon.IO;
using ENode.Commanding;
using Payments.Commands;
using Registration.Commands;
using Registration.Commands.Orders;
using Registration.ReadModel;
using Registration.Web.Extensions;
using Registration.Web.Models;
using Registration.Web.Utils;
using OrderStatus = Registration.Orders.OrderStatus;

namespace Registration.Web.Controllers
{
    public class RegistrationController : ConferenceTenantController
    {
        public const string ThirdPartyProcessorPayment = "thirdParty";
        private static readonly TimeSpan DraftOrderWaitTimeout = TimeSpan.FromSeconds(5);
        private static readonly TimeSpan DraftOrderPollInterval = TimeSpan.FromMilliseconds(750);
        private static readonly TimeSpan PricedOrderWaitTimeout = TimeSpan.FromSeconds(5);
        private static readonly TimeSpan PricedOrderPollInterval = TimeSpan.FromMilliseconds(750);

        private readonly ICommandService _commandService;

        public RegistrationController(ICommandService commandService, IConferenceQueryService conferenceQueryService, IOrderQueryService orderQueryService)
            : base(conferenceQueryService, orderQueryService)
        {
            _commandService = commandService;
        }

        [HttpGet]
        [OutputCache(Duration = 0, NoStore = true)]
        public ActionResult StartRegistration()
        {
            return View(CreateViewModel());
        }

        [HttpPost]
        public async Task<ActionResult> StartRegistration(OrderViewModel model)
        {
            if (!ModelState.IsValid)
            {
                return View(CreateViewModel());
            }

            var command = model.ToPlaceOrderCommand(this.ConferenceAlias, ConferenceQueryService);
            if (!command.Seats.Any())
            {
                ModelState.AddModelError("ConferenceCode", "You must reservation at least one seat.");
                return View(CreateViewModel());
            }

            var result = await SendCommandAsync(command);

            if (!result.IsSuccess())
            {
                ModelState.AddModelError("ConferenceCode", result.GetErrorMessage());
                return View(CreateViewModel());
            }

            return RedirectToAction("SpecifyRegistrantAndPaymentDetails", new { conferenceCode = this.ConferenceCode, orderId = command.AggregateRootId });
        }

        [HttpGet]
        [OutputCache(Duration = 0, NoStore = true)]
        public ActionResult SpecifyRegistrantAndPaymentDetails(Guid orderId)
        {
            var order = this.WaitUntilReservationCompleted(orderId).Result;
            if (order == null)
            {
                return View("PricedOrderUnknown");
            }

            if (order.Status == (int)OrderStatus.ReservationSuccess)
            {
                return View(new RegistrationViewModel
                {
                    RegistrantDetails = new RegistrantDetails { OrderId = order.OrderId },
                    Order = order
                });
            }
            else
            {
                return View("ReservationFailed");
            }
        }

        [HttpPost]
        public ActionResult SpecifyRegistrantAndPaymentDetails(Guid orderId, RegistrantDetails model, string paymentType)
        {
            if (!ModelState.IsValid)
            {
                return SpecifyRegistrantAndPaymentDetails(orderId);
            }

            SendCommandAsync(new AssignRegistrantDetails(orderId)
            {
                FirstName = model.FirstName,
                LastName = model.LastName,
                Email = model.Email
            });

            return this.StartPayment(orderId);
        }

        [HttpPost]
        public ActionResult StartPayment(Guid orderId)
        {
            var order = this.OrderQueryService.FindOrder(orderId);

            if (order == null)
            {
                return View("ReservationUnknown");
            }
            if (order.Status == (int)OrderStatus.PaymentSuccess || order.Status == (int)OrderStatus.Success)
            {
                return View("ShowCompletedOrder");
            }
            if (order.ReservationExpirationDate.HasValue && order.ReservationExpirationDate < DateTime.UtcNow)
            {
                return RedirectToAction("ShowExpiredOrder", new { conferenceCode = this.ConferenceAlias.Slug, orderId = orderId });
            }
            if (order.IsFreeOfCharge())
            {
                return CompleteRegistrationWithoutPayment(orderId);
            }

            return CompleteRegistrationWithThirdPartyProcessorPayment(order);
        }

        [HttpGet]
        [OutputCache(Duration = 0, NoStore = true)]
        public ActionResult ShowRejectedOrder(Guid orderId)
        {
            return View();
        }

        [HttpGet]
        [OutputCache(Duration = 0, NoStore = true)]
        public ActionResult ShowExpiredOrder(Guid orderId)
        {
            return View();
        }

        [HttpGet]
        [OutputCache(Duration = 0, NoStore = true)]
        public ActionResult ThankYou(Guid orderId)
        {
            return View(this.OrderQueryService.FindOrder(orderId));
        }

        private ActionResult CompleteRegistrationWithThirdPartyProcessorPayment(Order order)
        {
            var paymentCommand = CreatePaymentCommand(order);

            SendCommandAsync(paymentCommand);

            var paymentAcceptedUrl = this.Url.Action("ThankYou", new { conferenceCode = this.ConferenceAlias.Slug, order.OrderId });
            var paymentRejectedUrl = this.Url.Action("ShowRejectedOrder", new { conferenceCode = this.ConferenceAlias.Slug, orderId = order.OrderId });

            return RedirectToAction(
                "ThirdPartyProcessorPayment",
                "Payment",
                new
                {
                    conferenceCode = this.ConferenceAlias.Slug,
                    paymentId = paymentCommand.AggregateRootId,
                    paymentAcceptedUrl,
                    paymentRejectedUrl
                });
        }
        private CreatePayment CreatePaymentCommand(Order order)
        {
            var description = "Payment for the order of " + this.ConferenceAlias.Name;
            var paymentCommand = new CreatePayment
            {
                AggregateRootId = GuidUtil.NewSequentialId(),
                ConferenceId = this.ConferenceAlias.Id,
                OrderId = order.OrderId,
                Description = description,
                TotalAmount = order.TotalAmount,
                Lines = order.GetLines().Select(x => new PaymentLine { Id = x.SeatTypeId, Description = x.SeatTypeName, Amount = x.LineTotal })
            };

            return paymentCommand;
        }
        private ActionResult CompleteRegistrationWithoutPayment(Guid orderId)
        {
            SendCommandAsync(new MarkAsSuccess(orderId));
            return RedirectToAction("ThankYou", new { conferenceCode = this.ConferenceAlias.Slug, orderId });
        }
        private OrderViewModel CreateViewModel()
        {
            var seatTypes = this.ConferenceQueryService.GetPublishedSeatTypes(this.ConferenceAlias.Id);
            var viewModel = new OrderViewModel
            {
                ConferenceId = this.ConferenceAlias.Id,
                ConferenceCode = this.ConferenceAlias.Slug,
                ConferenceName = this.ConferenceAlias.Name,
                Items = seatTypes.Select(x => new OrderItemViewModel
                {
                    SeatType = x,
                    MaxSelectionQuantity = Math.Max(Math.Min(x.AvailableQuantity, 20), 0)
                }).ToList()
            };

            return viewModel;
        }

        /// <summary>轮训订单状态，直到订单的库存预扣操作完成
        /// </summary>
        /// <param name="orderId"></param>
        /// <returns></returns>
        private Task<Order> WaitUntilReservationCompleted(Guid orderId)
        {
            return TimerTaskFactory.StartNew<Order>(
                    () => this.OrderQueryService.FindOrder(orderId),
                    x => x != null && (x.Status == (int)OrderStatus.ReservationSuccess || x.Status == (int)OrderStatus.ReservationFailed),
                    PricedOrderPollInterval,
                    PricedOrderWaitTimeout);
        }
        /// <summary>异步发送给定的命令
        /// </summary>
        /// <param name="command"></param>
        /// <param name="millisecondsDelay"></param>
        /// <returns></returns>
        private Task<AsyncTaskResult> SendCommandAsync(ICommand command, int millisecondsDelay = 5000)
        {
            return _commandService.SendAsync(command).TimeoutAfter(millisecondsDelay);
        }
    }
}

File: samples/conference/registration-bc/src/main/java/com/microsoft/conference/registration/controller/RegistrationController.java
package com.microsoft.conference.registration.controller;

import com.microsoft.conference.common.ActionResult;
import com.microsoft.conference.common.ErrCode;
import com.microsoft.conference.common.payment.commands.CreatePayment;
import com.microsoft.conference.common.registration.commands.order.AssignRegistrantDetails;
import com.microsoft.conference.common.registration.commands.order.MarkAsSuccess;
import com.microsoft.conference.common.registration.commands.order.PlaceOrder;
import com.microsoft.conference.registration.domain.order.model.OrderStatus;
import com.microsoft.conference.registration.readmodel.PayConvert;
import com.microsoft.conference.registration.readmodel.RegistrationConvert;
import com.microsoft.conference.registration.readmodel.service.ConferenceAlias;
import com.microsoft.conference.registration.readmodel.service.ConferenceQueryService;
import com.microsoft.conference.registration.readmodel.service.OrderQueryService;
import com.microsoft.conference.registration.readmodel.service.OrderVO;
import com.microsoft.conference.registration.readmodel.service.SeatTypeVO;
import org.enodeframework.commanding.CommandResult;
import org.enodeframework.commanding.CommandReturnType;
import org.enodeframework.commanding.CommandStatus;
import org.enodeframework.commanding.ICommand;
import org.enodeframework.commanding.ICommandService;
import org.enodeframework.common.io.Task;
import org.enodeframework.common.utilities.IdGenerator;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.Date;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Collectors;

import static org.enodeframework.common.io.Task.await;

@RestController
public class RegistrationController {

    public String thirdPartyProcessorPayment = "thirdParty";

    @Autowired
    private ICommandService commandService;
    @Autowired
    private ConferenceQueryService conferenceQueryService;
    @Autowired
    private OrderQueryService orderQueryService;

    @GetMapping("registration/{id}")
    public ActionResult<OrderViewModel> startRegistration(@PathVariable String slug) {
        ConferenceAlias alias = conferenceQueryService.getConferenceAlias(slug);
        return view(createViewModel(alias));
    }

    @PostMapping("registration")
    public ActionResult<String> startRegistration(@RequestBody OrderViewModel model) {
        ConferenceAlias alias = conferenceQueryService.getConferenceAlias(model.getConferenceCode());
        PlaceOrder command = RegistrationConvert.INSTANCE.toPlaceOrderCommand(alias);
        if (!command.getSeatInfos().stream().findAny().isPresent()) {
            String errMsg = "ConferenceCode, You must reservation at least one seat.";
            return ActionResult.error(ErrCode.SYSTEM_ERROR, errMsg);
        }
        CommandResult result = await(executeCommandAsync(command));
        return view(result.getResult());
    }

    @GetMapping("specify")
    public ActionResult<RegistrationViewModel> specifyRegistrantAndPaymentDetails(@RequestParam String orderId) {
        OrderVO order = this.waitUntilReservationCompleted(orderId);
        if (order == null) {
            return ActionResult.error(ErrCode.SYSTEM_ERROR, "PricedOrderUnknown");
        }
        if (order.getStatus() != OrderStatus.ReservationSuccess.getStatus()) {
            return ActionResult.error(ErrCode.SYSTEM_ERROR, "ReservationFailed");
        }
        RegistrationViewModel viewModel = new RegistrationViewModel();
        viewModel.setOrderVO(order);
        RegistrantDetails registrantDetails = new RegistrantDetails();
        registrantDetails.setOrderId(orderId);
        viewModel.setRegistrantDetails(registrantDetails);
        return view(viewModel);
    }

    @PostMapping("specify")
    public ActionResult<String> specifyRegistrantAndPaymentDetails(String orderId, RegistrantDetails viewModel) {
        AssignRegistrantDetails registrantDetails = RegistrationConvert.INSTANCE.toAssignRegistrantDetails(viewModel);
        registrantDetails.setAggregateRootId(orderId);
        await(sendCommandAsync(registrantDetails));
        return this.startPayment(orderId);
    }

    @PostMapping("startpay")
    public ActionResult<String> startPayment(@RequestParam String orderId) {
        OrderVO order = this.orderQueryService.findOrder(orderId);
        if (order == null) {
            return ActionResult.error(ErrCode.SYSTEM_ERROR, "ReservationUnknown");
        }
        if (order.getStatus() == OrderStatus.PaymentSuccess.getStatus() || order.getStatus() == OrderStatus.Success.getStatus()) {
            return view("ShowCompletedOrder");
        }
        if (new Date().before(order.getReservationExpirationDate())) {
            return view("ShowExpiredOrder: " + orderId);
        }
        if (order.isFreeOfCharge()) {
            completeRegistrationWithoutPayment(orderId);
            return view("completeRegistrationWithoutPayment");
        }
        completeRegistrationWithThirdPartyProcessorPayment(order);
        return view("completeRegistrationWithThirdPartyProcessorPayment");
    }


    public ActionResult<OrderVO> thankYou(String orderId) {
        return view(this.orderQueryService.findOrder(orderId));
    }

    private void completeRegistrationWithThirdPartyProcessorPayment(OrderVO order) {
        CreatePayment paymentCommand = createPaymentCommand(order);
        sendCommandAsync(paymentCommand);
    }

    private CreatePayment createPaymentCommand(OrderVO order) {
        ConferenceAlias alias = conferenceQueryService.getConferenceAlias(order.getAccessCode());
        String name = alias.getName();
        String description = "Payment for the order of " + name;
        CreatePayment createPayment = new CreatePayment();
        createPayment.setAggregateRootId(IdGenerator.nextId());
        createPayment.setConferenceId(alias.getId());
        createPayment.setOrderId(order.getOrderId());
        createPayment.setDescription(description);
        createPayment.setTotalAmount(order.getTotalAmount());
        createPayment.setLines(order.getOrderLines().stream()
                .map(PayConvert.INSTANCE::toPaymentLine)
                .collect(Collectors.toList()));
        return createPayment;
    }

    private void completeRegistrationWithoutPayment(String orderId) {
        Task.await(sendCommandAsync(new MarkAsSuccess(orderId)));
    }

    private OrderViewModel createViewModel(ConferenceAlias alias) {
        // thread local
        List<SeatTypeVO> seatTypes = this.conferenceQueryService.getPublishedSeatTypes(alias.getId());
        OrderViewModel viewModel = new OrderViewModel();
        viewModel.setConferenceCode(alias.getSlug());
        viewModel.setConferenceId(alias.getId());
        viewModel.setConferenceName(alias.getName());
        viewModel.setItems(seatTypes.stream().map(x -> {
            OrderItemViewModel itemViewModel = new OrderItemViewModel();
            itemViewModel.setSeatTypeVO(x);
            itemViewModel.setMaxSelectionQuantity(Math.max(Math.min(x.getAvailableQuantity(), 20), 0));
            return itemViewModel;
        }).collect(Collectors.toList()));
        return viewModel;
    }

    // 轮训订单状态，直到订单的库存预扣操作完成
    private OrderVO waitUntilReservationCompleted(String orderId) {
        OrderVO x = this.orderQueryService.findOrder(orderId);
        if (x != null) {
            if (x.getStatus() == OrderStatus.ReservationSuccess.getStatus()
                    || x.getStatus() == OrderStatus.ReservationFailed.getStatus()) {
                return x;
            }
        }
        return null;
    }

    private CompletableFuture<Boolean> sendCommandAsync(ICommand command) {
        return commandService.sendAsync(command);
    }

    private boolean isSuceess(CommandResult result) {
        return CommandStatus.Success.equals(result.getStatus());
    }

    private CompletableFuture<CommandResult> executeCommandAsync(ICommand command) {
        return commandService.executeAsync(command, CommandReturnType.CommandExecuted);
    }

    private <T> ActionResult<T> view(T objects) {
        return new ActionResult<>(objects);
    }
}
